### PR TITLE
Replace JAI with ImageIO

### PR DESCRIPTION
**Pull request by twogee**
Undeprecate Image task in Java 9+. No documentation yet; only the code.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Windows/67/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Linux/61/

**Comment by bodewig** [comment]
I'm a bit torn as this PR breaks public APIs and may break build environments where somebody has managed to install JAI. for example because they've downloaded it long ago. Maybe it would be better to make this a new task so people can actively chose the implementation.

Also note we have deprecated the Java classes but the image task's manual doesn't mention deprecation at all.

**Comment by twogee** [comment]
JAI is dead. It depends on internal classes that are no longer there in Java 9. I propose dropping JAI API support completely, just like apt or javah. I don't like the like the idea of shipping dead code for the sake of BWC. If someone would like to use JAI on Java 8, there is an option of using an older ant-jai.jar; else, [jai-imageio-core](https://jai-imageio.github.io/jai-imageio-core/) implements the lacking functionality (TIFF support). Also, there are [TwelveMonkeys](https://haraldk.github.io/TwelveMonkeys/) for even more functionality.

The task is not deprecated in documentation because it works on Java 8 as is; perhaps the situation with JAI should be explained once there's an agreement on how to proceed.

**Comment by bodewig** [comment]
My scenario is people using Java 8, where JAI still works - Java 8 is far less dead than Java 9 in my view :-)

I'd prefer a new "imageio" (or whatever it is called) task that doesn't suggest it was a drop-in replacement.

**Comment by twogee** [comment]
"imageio" is fine by me, then we skip a potential debacle about naming of java-jai.jar 😉
Whenever JAI is laid to rest, "image" can be reclaimed as an alias.
FWIW Java 8 is an LTS release, but hopefully not the last one.

So, it's refactoring, fixing build.xml, writing the documentation and putting a note about JAI in Image task manual -- did I miss something?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Windows/68/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Linux/62/

File: build.xml [hunk 1]
```diff
@@ -277,13 +277,20 @@
...
+  </selector>
+
   <selector id="needs.jmf">
     <filename name="${optional.package}/sound/"/>
   </selector>
 
   <selector id="needs.jai">
     <or>
-      <filename name="${optional.package}/image/"/>
+      <filename name="${optional.package}/image/Image.*"/>
```
Comment (bodewig) [inline build.xml]: You also need to add ImageTest here as the selector is also used when deciding which tests to run (this probabl explains the Jenkins build failures).

**Comment by bodewig** [comment]
Looks as if only the manual was missing. :-)

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Windows/69/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Linux/63/

File: build.xml (hunk 1 above)
Comment (twogee) [reply build.xml]: Thanks, Stefan, on to the documentation...

**Comment by twogee** [comment]
I could almost reuse the class diagram in the documentation if it wasn't referring to PlanarImage... so I have to draw a new one unless the original UML is available somewhere (a long shot, I know...)

**Comment by bodewig** [comment]
or live without any UML diagram :-) - no, I don't think we've got anything but the image.

**Comment by twogee** [comment]
I decided to redraw the class diagram as GraphML (continuing Ivy tradition of using yEd), because it's quite informative.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Windows/72/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Linux/66/

**Comment by twogee** [comment]
Looking at the class diagram, I wonder why `addRectangle()` is in TransformOperation rather than only in Draw?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Linux/67/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Windows/73/

**Comment by bodewig** [comment]
Looks good to me. I probably wouldn't have touched the image task's diagram at all, but that's your choice.

No idea about `addRectangle`. If things work if you move the method over to `Draw` I suggest you do so for the imageio task and leave things as they are in image.

**Comment by twogee** [comment]
I will do move `addRectangle` and `addText` to `Draw`, it makes no sense to have them in an abstract class. The manual needs correction regarding `ImageOperation`, both for image and imageio tasks: it should not be mentioned at all. Even if imageoperation could be used as a nested tag, it will only cause instantiation exception. Besides, text has a wrong attribute, it's "string", not "text"; so it makes sense to redraw both class diagrams 😉

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Windows/74/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Linux/68/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Windows/75/

**Comment by bodewig** [comment]
`addImageOperation` indeed looks strange (as does the manual when it talks about `ImageOperation`, By virtue of `add(ImageOperation)` you can add any type as nested element that implements `ImgeOperation` which is a good thing.

Not sure what you mean with "text has the wrong attribute".

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Linux/71/

**Comment by twogee** [comment]
Thanks for reviewing. I hope my last refactoring clarifies the point about the string field in the Text class (or the string attribute in the text element).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Windows/77/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Windows/80/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Linux/74/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ant%20Github-PR-Linux/75/

**Pull request closed without merging**